src/content/contribute/writing-a-plugin.md

```diff
@@ -3,125 +3,141 @@ title: Writing a Plugin
 
-  // Setup callback for accessing a compilation:
-  compiler.plugin("compilation", function(compilation) {
+These two components are an integral part of any webpack plugin (especially a `compilation`), so developers will benefit by familiarizing themselves with these source files:
+
+- [Compiler Source](https://github.com/webpack/webpack/blob/master/lib/Compiler.js)
+- [Compilation Source](https://github.com/webpack/webpack/blob/master/lib/Compilation.js)
 
-    // Now setup callbacks for accessing compilation steps:
-    compilation.plugin("optimize", function() {
-      console.log("Assets are being optimized.");
+## Accessing Compilation
+
+Compiler exposes a bunch of hooks that provide a reference to each new compilation. Compilations, in their turn, provide additional event hooks for tapping into steps within the build process.
+
+```javascript
+class HelloCompilationPlugin {
+  apply(compiler) {
+    // Setup callback for accessing a compilation:
+    compiler.hooks.compilation.tap("HelloCompilationPlugin", (compilation) => {
+      // Now setup callbacks for accessing compilation steps:
+      compilation.hooks.optimize.tap("HelloCompilationPlugin", () => {
+        console.log("Hello compilation!");
+      });
     });
-  });
-};
+  }
+}
 
 module.exports = HelloCompilationPlugin;
 ```
 
-For more information on what callbacks are available on the `compiler`, `compilation`, and other important objects, see the [plugins](/api/plugins/) doc.
+The list of hooks available on the `compiler`, `compilation`, and other important objects, see the [plugins API](/api/plugins/) docs.
 
-## Async compilation plugins
+## Async event hooks
 
-Some compilation plugin steps are asynchronous, and pass a callback function that _must_ be invoked when your plugin is finished running.
+Some event hooks are asynchronous. Apart from `tap`, they also have `tapAsync` and `tapPromise` methods. By tapping using these methods, you can do asynchronous actions inside hooks:
 
 ```javascript
-function HelloAsyncPlugin(options) {}
-
-HelloAsyncPlugin.prototype.apply = function(compiler) {
-  compiler.plugin("emit", function(compilation, callback) {
+class HelloAsyncPlugin {
+  apply(compiler) {
+    // tapAsync() is callback-based
+    compiler.hooks.emit.tapAsync('HelloAsyncPlugin', function(compilation, callback) {
+      // Do something async...
+      setTimeout(function() {
+        console.log("Done with async work...");
+        callback();
+      }, 1000);
+    });
 
-    // Do something async...
-    setTimeout(function() {
-      console.log("Done with async work...");
-      callback();
-    }, 1000);
+    // tapPromise() is promise-based
+    compiler.hooks.emit.tapPromise('HelloAsyncPlugin', (compilation) => {
+      // Do something async...
+      return doSomethingAsync()
```

Comment:
The comment from above feels redundant, we can put something like _Task that will run asynchronously_, or I will remove it, I think the context gives enough explanation.